torch/csrc/jit/codegen/cuda/lower_thread_predicate.cpp

```diff
@@ -0,0 +1,194 @@
+
+  return pred->as<Bool>();
+}
+
+} // namespace
+
+std::bitset<num_p_type> ThreadPredicates::getThreadPredicates(
+    const TensorView* tv) {
+  TORCH_INTERNAL_ASSERT(
+      thread_predicates.find(tv) != thread_predicates.end(),
+      "Invalid predicate initialization, couldn't find ",
+      tv);
+  return thread_predicates[tv];
+}
+
+// Update the reduction_deps bitset based on provided Expr
+void ThreadPredicates::updateBitSet(Expr* expr) {
+  // Which predicates were set for the inputs
+  std::bitset<num_p_type> input_preds;
+
+  // Which dims are reductions in inputs
+  std::bitset<num_p_type> input_reductions;
+
+  // Which dims are bcast in inputs
+  std::bitset<num_p_type> input_bcasts;
+
+  // Run through inputs and update bitsets
+  for (const auto* inp : expr->inputs()) {
+    if (!ir_utils::isTV(inp))
+      continue;
+
+    auto tv_inp = ir_utils::asConstTV(inp);
+    TORCH_INTERNAL_ASSERT(
+        thread_predicates.find(tv_inp) != thread_predicates.end(),
+        "Thread predicate map was not initialized, couldn't find ",
+        inp);
+
+    input_preds |= thread_predicates[tv_inp];
+
+    std::bitset<num_p_type> id_reductions;
+    std::bitset<num_p_type> id_bcasts;
+    std::bitset<num_p_type> id_ptypes;
+
+    for (auto id : tv_inp->domain()->domain()) {
+      if (id->isThread()) {
+        flip_true(id_ptypes, id->parallel_method());
+        if (id->isReduction())
+          flip_true(id_reductions, id->parallel_method());
+        if (id->isBroadcast())
+          flip_true(id_bcasts, id->parallel_method());
+      }
+    }
+
+    // Validate the combination of ptypes, reductions, bcasts
+    for (size_t i = 0; i < num_p_type; i++) {
+      if (input_reductions[i]) {
+        if (id_ptypes[i]) {
+          TORCH_INTERNAL_ASSERT(
+              id_reductions[i],
+              "Mismatched parallelized reductions found on inputs of epxr: ",
+              expr);
+          TORCH_CHECK(
+              !id_bcasts[i],
+              "Invalid broadcast and reduction combination, tried to parallelize both with the same thread dim: ",
+              inp);
+        }
+      } else if (input_bcasts[i]) {
+        if (id_ptypes[i])
+          TORCH_INTERNAL_ASSERT(
+              id_bcasts[i],
+              "Mismatched parallelized broadcast found on inputs of epxr: ",
+              expr,
+              ".");
+      }
+    }
+
+    // Accumulate
+    input_reductions |= id_reductions;
+    input_bcasts |= id_bcasts;
+  }
+
+  // Update map for this tv, before accumulating to other inputs
+  // Add any reductions this id has to any input predicates
+  auto output_preds = input_preds | input_reductions;
+
+  // Figure out which dims bcast wants to reset
+  auto bcast_reset_map = output_preds & input_bcasts;
+
+  // Flip it to make a bit mask
+  bcast_reset_map = ~bcast_reset_map;
+
+  // Get rid of any reductions which are bcasted
+  output_preds &= bcast_reset_map;
```

Comment:
... next step is to make output_preds const too:

const auto output_preds = (input_preds | input_reduction) & reset_map;